compiler/shared/main/scala/mlscript/compiler/ir/IR.scala

```diff
@@ -0,0 +1,361 @@
+package mlscript.compiler.ir
+
+import mlscript._
+import mlscript.utils._
+import mlscript.utils.shorthands._
+import mlscript.compiler.ir._
+import mlscript.compiler.optimizer._
+
+import collection.mutable.{Map as MutMap, Set as MutSet, HashMap, ListBuffer}
+import annotation.unused
+import util.Sorting
+import scala.collection.immutable.SortedSet
+
+final case class IRError(message: String) extends Exception(message)
+
+case class Program(
+  val classes: Set[ClassInfo],
+  val defs: Set[Defn],
+  val main: Node,
+):
+  override def equals(o: Any): Bool = o match {
+    case o: Program if this.isInstanceOf[Program] =>
+      o.classes == classes &&
+      o.defs == defs &&
+      o.main == main
+    case _ => false
+  }
+  override def toString: String =
+    val t1 = classes.toArray
+    val t2 = defs.toArray
+    Sorting.quickSort(t1)
+    Sorting.quickSort(t2)
+    s"Program({${t1.mkString(",")}}, {\n${t2.mkString("\n")}\n},\n$main)"
+
+implicit object ClassInfoOrdering extends Ordering[ClassInfo] {
+  def compare(a: ClassInfo, b: ClassInfo) = a.id.compare(b.id)
+}
+
+case class ClassInfo(
+  val id: Int,
+  val ident: Str,
+  val fields: Ls[Str],
+):
+  override def equals(o: Any): Bool = o match {
+    case o: ClassInfo if this.isInstanceOf[ClassInfo] => o.id == id
+    case _ => false
+  }
+  override def hashCode: Int = id
+  override def toString: String =
+    s"ClassInfo($id, $ident, [${fields mkString ","}])"
+
+case class Name(val str: Str):
+  override def equals(o: Any): Bool = o match {
+    case Name(s2) => str == s2
+    case _ => false
+  }
+
+  def copy: Name = Name(str)
+  def trySubst(map: Map[Str, Name]) = map.getOrElse(str, this)
+  
+  override def toString: String =
+    str
+
+class DefnRef(var defn: Either[Defn, Str]):
+  def getName: String = defn match {
+    case Left(godef) => godef.getName
+    case Right(name) => name
+  }
+  def expectDefn: Defn = defn match {
+    case Left(godef) => godef
+    case Right(name) => throw Exception(s"Expected a def, but got $name")
+  }
+  
+  def getDefn: Opt[Defn] = defn match {
+    case Left(godef) => Some(godef)
+    case Right(name) => None
+  }
+
+  override def equals(o: Any): Bool = o match {
+    case o: DefnRef if this.isInstanceOf[DefnRef] =>
+      o.getName == this.getName
+    case _ => false
+  }
+
+
+implicit object DefOrdering extends Ordering[Defn] {
+  def compare(a: Defn, b: Defn) = a.id.compare(b.id)
+}
+
+case class Defn(
+  val id: Int,
+  val name: Str,
+  val params: Ls[Name],
+  val resultNum: Int,
+  var specialized: Opt[Ls[Opt[Intro]]],
+  val body: Node
+):
+
+  override def equals(o: Any): Bool = o match {
+    case o: Defn if this.isInstanceOf[Defn] =>
+      o.id == id &&
+      o.body == body
```

Comment:
Why do you need to check `body` here? Do you? It's an expensive check; should be avoided.